### ISSUE TITLE
JavaScript - Pinscape NightMode Event, NightMode Property

### ISSUE THREAD

**Issue by Zappo-II** [mjrgh/PinballY#258]
**Is your feature request related to a problem? Please describe.**
I tried to find a way to obtain the current nightmode state of pinscape connected to the cabinet but found none...
I think the information is present inside PinballY since it visually checks / unchecks the NightMode entry in the operator menu...

**Describe the solution you'd like**
I would really appreciate to have a Property like system.nightmodeavailable boolen (if pinscape present true else false) and system.nightmodestate boolean (night mode on := true, else := false)...
Additionally it would be great to have an event like NightModeChanged(), that gets triggered if night mode changed by ANY source and gives the information of the current state...

**Describe alternatives you've considered**
I think I could obtain the State Information by triggering external program PinscapeConfigToolCLI.exe bur that would be overkill in my opinion, since the information is already available inside PinballY. And on the other hand that would have to be scheduled sampling and could loose some of the changes in between...

**Additional context**
I am about to build a HomeAssistant integration (Signaling if Table is Playing or in Idle/Attract Mode, Displaying current BackGlass Picture, etc... - Already working perfectly well) and I would like to be able to signal the current night mode state there also...

If I somehowe have missed obvious (or even not so obvious) ways to obtain this in JavaScript without Feature / Change Requests, please advise...

**Comment by mjrgh** [comment]
You're right that there's no way to get the live Night Mode status from Javascript at the moment.  PinballY does have the information internally but it's not exposed through any JS interfaces.  I agree this would be great to have - I'll see what I can do.

**Comment by mjrgh** [comment]
Here's a quick proposal - let me know what you think.

New 'pinscape' global object:

pinscape.nightMode - read to get current Night Mode status (true -> Night Mode is enabled in at least one attached Pinscape unit, false -> all units are in "day mode", or no units are attached at all); set to true to enable Night Mode on all attached controllers, set to false to return all controllers to "day mode".
```
// set night mode
pinscape.nightMode = true;
```
PinballY doesn't have any internal native event related to Night Mode, since it doesn't monitor the Pinscape status continuously. I could pretty easily add an internal timer that polls the status and generates an event on the pinscape object (something like `pinscape.on("nightmode" => { ... })`), but I'm not sure there'd be any advantage to that, compared to just letting you do essentially the same thing with a setInterval() of your own.  Doing it with a user setInterval() would let you choose your own polling interval, to balance how quickly you need it to respond against the increased overhead of increased polling frequency.
```
let oldNightMode = pinscape.nightMode;
setInterval(() => {
    let newNightMode = pinscape.nightMode;
    if (newNightMode != oldNightMode) {
        // react to the change
        oldNightMode = newNightMode;
    }
}, 1000);  // check every 1000ms = every 1 second
```
It's not quite as compact as if there were a system event for it, so I guess that's a point in favor of making it a system event, but it's still pretty straightforward, and it gives you the flexibility to control the timing.

**Comment by Zappo-II** [comment]
Amazing... :vulcan_salute: ...

Did not think of the possibility to have more than one pinscape connected...
Have only one and no idea if one having more than one should be able to set their nightmode states independently... :thinking: ...

I, on the other hand, would be more than lucky if you could implement it as you proposed... :innocent: ...

If you are technically unable to implement a proper event then I would also suggest to leave it to the `setInterval` way...
The response time of the usb device poll behind the `let newNightMode = pinscape.nightMode;` should not take any significant amount of time, I'd assume, I'd also assume the PinballY internal update of the `pinscape.nightMode` property will take place when the property is read and thus be transient, won't it...???

**TL/DR;**

* Perfect - I'd take the globalObject.Property and stick to setInterval ...

**Comment by mjrgh** [comment]
> I'd also assume the PinballY internal update of the pinscape.nightMode property will take place when the property is read and thus be transient, won't it...???

Exactly - pinscape.nightMode would be a getter/setter, so nothing is actually stored locally, it's just a window into the live USB state.  Each time you read the property, you get a fresh, live result from the USB connection(s), and writing it immediately sends the USB commands to update the mode on connected devices.

It's not that it would be technically infeasible to add a formal event for Night Mode changes, it's more that I don't think there's any advantage to it compared to the setInterval approach, other than perhaps the more compact notation it would allow.  There's a slight downside to adding it on the C++ side, which is that if you're not running any scripts that depend on the event, the added overhead of the polling would be wasted work.  The setInterval approach only incurs the overhead if you're actually getting use out of it, plus it gives you control over how frequently you want the polling to be done (if it were built in, I'd just have to pick a one-size-fits-all polling rate).  I can always add a formal event in the future if the setInterval approach turns out in practice to have any downsides I'm not thinking of right now.

**Comment by Zappo-II** [comment]
Property and setInterval will do perfectly well from my point of view...

THX for getting into this, looking forward to getting your implementation... :innocent: ...
